### ISSUE TITLE
Slice hashing thoughts

### ISSUE THREAD

**Issue by patriknw** [akka/akka-persistence-r2dbc#175]
For the journal table we have a HASH partition key:
```
PRIMARY KEY((slice, entity_type) HASH, persistence_id, seq_nr ASC)
```

One potential problem is that we have rather few slices so there is a risk that the rows are not evenly distributed over the tablets (db sharding).

Given the table:
```
CREATE TABLE IF NOT EXISTS test_journal(
  slice INT NOT NULL,
  entity_type VARCHAR(255) NOT NULL,
  persistence_id VARCHAR(255) NOT NULL,
  seq_nr BIGINT NOT NULL,
  db_timestamp timestamp with time zone NOT NULL,
  event VARCHAR(255),

  PRIMARY KEY((slice, entity_type) HASH, persistence_id, seq_nr ASC)
);
```

and inserting 1 row per slice:
```
DO
$do$
BEGIN
  for i in 0..127 loop
    insert into test_journal (slice, entity_type, persistence_id, seq_nr, db_timestamp, event)
    values (i, 'MyEntity', 'pid-' || i, 1, transaction_timestamp(), 'event-' || i);
  end loop;
END
$do$;
```

Yugabyte is using 2 bytes for the shard hash code, i.e. the range is 0-65535. By default it slits a table in 8 tablets per node, i.e. 24 tablets per table so let's experiment with that value. That means that each tablet takes a range of 2730 hash values.

We can look at the distributions with the following sql:
```
select slice, yb_hash_code(slice::int, entity_type::VARCHAR(255)) as h, yb_hash_code(slice::int, entity_type::VARCHAR(255)) / 2730 as t from test_journal order by h;

select yb_hash_code(slice::int, entity_type::VARCHAR(255)) / 2730 as t, count(*) from test_journal group by t order by t;

create table slice_distribution as (select yb_hash_code(slice::int, entity_type::VARCHAR(255)) / 2730 as t, count(*) from test_journal group by t order by t);

select min(count), max(count), min(count)::decimal * 100 / sum(count) as min_percent, max(count)::decimal * 100 / sum(count) as max_percent from slice_distribution;

select * from slice_distribution order by count;
 t  | count
----+-------
  4 |     2
  6 |     2
 17 |     2
 23 |     3
  3 |     3
 15 |     4
  8 |     4
 13 |     4
 16 |     4
 19 |     4
 14 |     5
 12 |     5
 18 |     6
  0 |     6
 11 |     6
 22 |     6
 10 |     6
  9 |     7
  5 |     7
  7 |     8
  1 |     8
  2 |     8
 20 |     9
 21 |     9
```

That doesn't look great. Some get 2 of 128 slices, and some 9 of 128.

One can repeat the experiment with 1024 slices, and it becomes better but would that be good enough?

```
t  | count
----+-------
 23 |    34
  8 |    34
  4 |    35
 12 |    37
  6 |    38
 22 |    38
 14 |    39
 17 |    39
 10 |    40
 21 |    42
 15 |    42
 13 |    42
  9 |    43
 19 |    43
  1 |    44
  0 |    44
 11 |    45
  3 |    45
  2 |    49
  7 |    49
 16 |    49
  5 |    49
 20 |    51
 18 |    53
 ```
 
 ## Why slice as sharding key for journal table?
 
 We don't have to use the slice as sharding key in the journal table. We could use the `persistence_id` instead and that would give good distribution assuming that persistence_id is random.
 
 The slice is needed for the `eventsBySlices` query (Projections) but for that we have the slice index.
 
I started out with the slice as sharding key in the journal table because I thought it was a "natural" sharding key that we anyway had in place. I also found the "Co-partitioning" feature in https://github.com/yugabyte/yugabyte-db/blob/master/architecture/design/ysql-tablegroups.md which could make the index and the parent table colocated if they have the same sharding key. That feature doesn't exist yet.

`eventsBySlices` would search the index and then retrieve the event payload for matching entries from the parent journal table. My thinking was that such fetch would be most efficient if it would be in a batch request from one (a few) tablets rather than spread over all tablets. That's just speculation, though.

## Why use HASH sharding for journal table?

Most recommendations in the Yugabyte documentation prefer hash sharding but maybe range sharding would be equally good for the slice in the journal table.

**Comment by jrudolph** [comment]
I read the ticket yesterday but missed the point it seems :)

The point is that distributing 128 slices to 24 buckets will always lead to a bad distribution just due to statistics. I ran a few simulations and the one you posted above seems pretty standard.

Basically, we are doing two hashing steps here: persistence_id => slice, and slice => tablet. Each hashing step can be seen as a decrease in "resolution". It's a bit like reducing image resolution twice (or compressing it twice), the result will be less accurate than doing the full reduction in a single step.

So, I think we should indeed either hash on persistence_ids or significantly increase the number of slices (is there any downside to that?).

**Comment by patriknw** [comment]
Thanks for confirming.

**Comment by patriknw** [comment]
I have tried with using `persistence_id` as the sharding key in the journal:
```
PRIMARY KEY(persistence_id HASH, seq_nr ASC)
```

and also pre-split the slice index:
```
CREATE INDEX IF NOT EXISTS event_journal_slice_idx ON event_journal(slice ASC, entity_type ASC, db_timestamp ASC)
  SPLIT AT VALUES ((31), (63), (95));
```
(there seems to be a problem with combining INCLUDE and SPLIT AT, but that is another issue)

With the projection testing app it has a rough write rate of 3000 inserts/s. The projections are keeping up.

The eventsBySlices query looks like:
```
explain analyze SELECT slice, persistence_id, seq_nr, db_timestamp, statement_timestamp() AS read_db_timestamp, event_ser_id, event_ser_manifest, event_payload, meta_ser_id, meta_ser_manifest, meta_payload
  FROM event_journal WHERE slice BETWEEN 0 AND 31 AND entity_type = 'configurable'
  AND db_timestamp >= '2021-12-14 08:13:00.042089+00' AND db_timestamp < transaction_timestamp() - interval '200 milliseconds'
  AND deleted = false
  ORDER BY db_timestamp, seq_nr
  LIMIT 1000;

-----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=5.57..5.59 rows=10 width=1648) (actual time=80.071..80.239 rows=1000 loops=1)
   ->  Sort  (cost=5.57..5.59 rows=10 width=1648) (actual time=80.070..80.175 rows=1000 loops=1)
         Sort Key: db_timestamp, seq_nr
         Sort Method: top-N heapsort  Memory: 900kB
         ->  Index Scan using event_journal_slice_idx on event_journal  (cost=0.00..5.40 rows=10 width=1648) (actual time=9.007..75.905 rows=14351 loops=1)
               Index Cond: ((slice >= 0) AND (slice <= 31) AND ((entity_type)::text = 'configurable'::text) AND (db_timestamp >= '2021-12-14 08:13:00.042089+00'::timestamp with time zone) AND (db_timestamp < (transaction_timestamp() - '00:00:00.2'::interval)))
               Filter: (NOT deleted)
 Planning Time: 0.092 ms
 Execution Time: 80.460 ms
 ```
 That was 14000 rows from the tail. Looking further back (251000 rows from the tail) it increases (but that is probably not so important):
 ```
 explain analyze SELECT slice, persistence_id, seq_nr, db_timestamp, statement_timestamp() AS read_db_timestamp, event_ser_id, event_ser_manifest, event_payload, meta_ser_id, meta_ser_manifest, meta_payload
  FROM event_journal WHERE slice BETWEEN 0 AND 31 AND entity_type = 'configurable'
  AND db_timestamp >= '2021-12-14 08:00:00.042089+00' AND db_timestamp < transaction_timestamp() - interval '200 milliseconds'
  AND deleted = false
  ORDER BY db_timestamp, seq_nr
  LIMIT 1000;

----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=5.57..5.59 rows=10 width=1648) (actual time=1391.950..1392.119 rows=1000 loops=1)
   ->  Sort  (cost=5.57..5.59 rows=10 width=1648) (actual time=1391.948..1392.063 rows=1000 loops=1)
         Sort Key: db_timestamp, seq_nr
         Sort Method: top-N heapsort  Memory: 564kB
         ->  Index Scan using event_journal_slice_idx on event_journal  (cost=0.00..5.40 rows=10 width=1648) (actual time=9.526..1329.767 rows=251000 loops=1)
               Index Cond: ((slice >= 0) AND (slice <= 31) AND ((entity_type)::text = 'configurable'::text) AND (db_timestamp >= '2021-12-14 08:00:00.042089+00'::timestamp with time zone) AND (db_timestamp < (transaction_timestamp() - '00:00:00.2'::interval)))
               Filter: (NOT deleted)
 Planning Time: 0.097 ms
 Execution Time: 1392.401 ms
 ```
 
 That was with > 4 million rows in the event_journal.

**Comment by patriknw** [comment]
Tried with range sharding for the journal table:
```
PRIMARY KEY(slice ASC, entity_type ASC, persistence_id ASC, seq_nr ASC)

SPLIT AT VALUES ((7), (15), (23), (31), (39), (47), (55), (63), (71), (79), (87), (95), (103), (111), (121))
```

The rough insert rate for 1 million events is 2600 inserts/s, which is lower than the 3000 inserts/s in previous test.

`pg_stat_statements` shows slightly longer average time for the insert compared to previous test. 9.03 vs 8.49.

The eventsBySlices query 14000 rows from the tail:

```
explain analyze SELECT slice, persistence_id, seq_nr, db_timestamp, statement_timestamp() AS read_db_timestamp, event_ser_id, event_ser_manifest, event_payload, meta_ser_id, meta_ser_manifest, meta_payload
  FROM event_journal WHERE slice BETWEEN 0 AND 31 AND entity_type = 'configurable'
  AND db_timestamp >= '2021-12-14 11:51:37.042089+00' AND db_timestamp < transaction_timestamp() - interval '200 milliseconds'
  AND deleted = false
  ORDER BY db_timestamp, seq_nr
  LIMIT 1000;
                                                                                                                             QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=5.57..5.59 rows=10 width=1648) (actual time=468.389..468.557 rows=1000 loops=1)
   ->  Sort  (cost=5.57..5.59 rows=10 width=1648) (actual time=468.387..468.501 rows=1000 loops=1)
         Sort Key: db_timestamp, seq_nr
         Sort Method: top-N heapsort  Memory: 840kB
         ->  Index Scan using event_journal_slice_idx on event_journal  (cost=0.00..5.40 rows=10 width=1648) (actual time=38.400..460.857 rows=15130 loops=1)
               Index Cond: ((slice >= 0) AND (slice <= 31) AND ((entity_type)::text = 'configurable'::text) AND (db_timestamp >= '2021-12-14 11:51:37.042089+00'::timestamp with time zone) AND (db_timestamp < (transaction_timestamp() - '00:00:00.2'::interval)))
               Filter: (NOT deleted)
 Planning Time: 0.143 ms
 Execution Time: 469.327 ms
 ```
 
 250710 rows behind the tail:
 ```
 explain analyze SELECT slice, persistence_id, seq_nr, db_timestamp, statement_timestamp() AS read_db_timestamp, event_ser_id, event_ser_manifest, event_payload, meta_ser_id, meta_ser_manifest, meta_payload
  FROM event_journal WHERE slice BETWEEN 0 AND 31 AND entity_type = 'configurable'
  AND db_timestamp >= '2021-12-14 11:45:00.042089+00' AND db_timestamp < transaction_timestamp() - interval '200 milliseconds'
  AND deleted = false
  ORDER BY db_timestamp, seq_nr
  LIMIT 1000;
                                                                                                                             QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=5.57..5.59 rows=10 width=1648) (actual time=7558.359..7558.551 rows=1000 loops=1)
   ->  Sort  (cost=5.57..5.59 rows=10 width=1648) (actual time=7558.358..7558.493 rows=1000 loops=1)
         Sort Key: db_timestamp, seq_nr
         Sort Method: top-N heapsort  Memory: 564kB
         ->  Index Scan using event_journal_slice_idx on event_journal  (cost=0.00..5.40 rows=10 width=1648) (actual time=36.639..7498.044 rows=250710 loops=1)
               Index Cond: ((slice >= 0) AND (slice <= 31) AND ((entity_type)::text = 'configurable'::text) AND (db_timestamp >= '2021-12-14 11:45:00.042089+00'::timestamp with time zone) AND (db_timestamp < (transaction_timestamp() - '00:00:00.2'::interval)))
               Filter: (NOT deleted)
 Planning Time: 0.104 ms
 Execution Time: 7559.047 ms
 ```

Using same index definition here as in previous test
```
CREATE INDEX IF NOT EXISTS event_journal_slice_idx ON event_journal(slice ASC, entity_type ASC, db_timestamp ASC)
  SPLIT AT VALUES ((31), (63), (95));
```

In conclusion. That didn't work well. The explain cost numbers are the same, but the actual execution times are much higher.
